### PR TITLE
Use local labels in padlock.c

### LINKED ISSUE #3451: Use local labels for inline asm
### Description
- Type: Bug
- Priority: Blocker

---------------------------------------------------------------
## Bug

**OS**  
iPhoneSimulator

**mbed TLS build:**  
Version: 2.22.0
OS version: 9.0
Configuration: default
Compiler and options (if you used a pre-built binary, please indicate how you obtained it):  Xcode 11.5
Additional environment information:  

**Expected behavior**   
building success
**Actual behavior**  
got error when linking
```
<inline asm>:9:2: error: invalid symbol redefinition
        unsupported:              
        ^
LLVM ERROR: Error parsing inline asm

clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

`unsupported:` is located in `padlock.c` which use named labels in inline asm.
**Steps to reproduce**  
I can't make an reproducible minimized project.
but according https://stackoverflow.com/questions/14506151/invalid-symbol-redefinition-in-inline-asm-on-llvm.
we should use local labels for inline asm, not named labels.

### PR DESCRIPTION
**Pull request by okhowang**
## Description
fix #3451 

## Status
READY

## Requires Backporting
NO

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

## Steps to test or reproduce
Outline the steps to test or reproduce the PR here.

### CODE REVIEW

**Comment by okhowang** [comment]
done

File: ChangeLog.d/bugfix_PR3452.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Bugfix
+   * Use local labels for fixing inline asm invalid symbol redefinition
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/bugfix_PR3452.txt:2]: It isn't clear under what circumstances the error happens. Given the nature of the change, I think an accurate description would be:

> Use local labels in mbedtls_padlock_has_support() to fix an invalid symbol redefinition if the function is inlined.

**Comment by gilles-peskine-arm** [comment]
The PR description says “needs backports” and “backported”, but no backports have been submitted. Are backports actually necessary here? Is there a risk that the change might not work in some environments (e.g. with older compilers or with toolchains other than GNU and Clang)?

**Comment by okhowang** [comment]
no backport please.
I have changed description.

**Comment by okhowang** [comment]
Changelog changed on demand.

**Comment by mpg** [comment]
This fixes #3451 which is a failure to build in some cases, hence a bug, so I'm relabeling this from "enhancement" to "bug". (Also, the ChangeLog entry in this PR already says "bugfix", so let's be consistent.)

As a general rule, bug fixes should be backported unless there's a specific reason not to. Do we have a reason not to backport this bug fix?

**Comment by mpg** [comment]
I'm labeling "needs: backports" not to say that we need backports, but as a reminder that if we don't need them, we want to document why.

**Comment by okhowang** [comment]
symbol redefinition error is occurred in a condition which is unknown currently.
and it was reported by only me so far.
can we regard it as a bug in a very rare situation?
maybe there is no need to backport?

**Comment by gilles-peskine-arm** [comment]
My analysis on backporting: this is a platform incompatibility, so if it's already working for someone, there's a good chance that it'll keep working. This is a weak argument because it could stop working due to a compiler upgrade or, given the nature of the issue, due to a code change that affects how the function is inlined. On the other hand, there is a risk that some toolchains might not support the local label syntax, but again this is a weak argument since it seems that [at least most Unixy assemblers do](https://stackoverflow.com/questions/3898435/labels-in-gcc-inline-assembly/3898451#comment30997301_3898451). As far as I understand, the problem arises if `mbedtls_padlock_has_support` gets inlined, but that can only happen with cross-module inlining, and realistically it'll only happen if both `mbedtls_aes_setkey_enc` and `mbedtls_aes_setkey_dec` are called from the same function (or functions inlined into the same function) and are both inlined, which is fairly rare and we haven't been able to reproduce so far. Arguably, it's a compiler bug, since the compiler shouldn't try to inline the function if it's not capable of doing so.

So I come out pretty close to the middle: there's some low value to backporting, and some low risk. The risk seems small, so I'm weakly in favor, but I'm fine with not backporting.

That being said, backporting would take less time than this discussion is shaping up to take.

**Comment by mpg** [comment]
Ok, so let's save time by not discussing this further and not backporting. At least now the reasons why we didn't backport are documented in case we (or someone else) wonder in the future why this wasn't backported.

**Comment by akien-mga** [comment]
Hello! Is this something that could be considered for a backport to 2.16 LTS?

We've been disabling padlock for a while in Godot due to this conflict, and there's no issue for us to keep patching it (we'll likely backport this patch to replace our own), but if this gets upstreamed in a future 2.16 release that's even better :)

**Comment by gilles-peskine-arm** [comment]
@akien-mga Mbed TLS 2.16 LTS has reached the end of its 3-year minimum support period and we are not extending it. As [announced](https://lists.trustedfirmware.org/pipermail/mbed-tls/2021-December/000508.html), last week's 2.16.12 release is the last in the 2.16 series.

Please update to the new 2.28 LTS.

**Comment by Faless** [comment]
@gilles-peskine-arm thank you! I had missed that message, it would be worth updating the [2.28.0 release notes](https://github.com/ARMmbed/mbedtls/releases/tag/v2.28.0) to mention that this is the new LTS release.
I did read that 2.16 TLS support was "until end of december 2021" but couldn't find which was supposed to supersede it as LTS release. Again, thank you for your help and your work!

**Pull request merged**